sdk/storage/azure-storage-blob/azure/storage/blob/_shared/uploads_async.py

```diff
@@ -421,3 +423,51 @@ async def _upload_chunk(self, chunk_offset, chunk_data):
     # TODO: Implement this method.
     async def _upload_substream_block(self, index, block_stream):
         pass
+
+
+class AsyncIterStreamer():
+    """
+    File-like streaming object for AsyncGenerators.
+    """
+    def __init__(self, generator: AsyncGenerator[Union[bytes, str], None], encoding: str = "UTF-8"):
+        self.generator = generator
+        self.iterator = generator.__aiter__()
+        self.leftover = b""
+        self.encoding = encoding
+
+    def __len__(self):
+        return self.generator.__len__()
+
+    def __aiter__(self):
+        return self.iterator
+
+    async def __anext__(self):
+        return await self.iterator.__anext__()
+
+    def seekable(self):
+        return False
+
+    def tell(self, *args, **kwargs):
+        raise UnsupportedOperation("Data generator does not support tell.")
+
+    def seek(self, *args, **kwargs):
+        raise UnsupportedOperation("Data generator is not seekable.")
+
+    async def read(self, size: int) -> bytes:
+        data = self.leftover
+        count = len(self.leftover)
+        try:
+            while count < size:
+                chunk = await self.__anext__()
+                if isinstance(chunk, str):
+                    chunk = chunk.encode(self.encoding)
+                data += chunk
+                count += len(chunk)
+        # This means count < size and what's leftover will be returned in this call.
+        except StopAsyncIteration:
+            self.leftover = b""
+
+        if count >= size:
+            self.leftover = data[size:]
```

Comment:
If I do an upload with `concurrency > 1` am I going to hit race conditions around this?

Reply:
That's a good question. This is the same exact logic as the sync version of this object which we have never had a report of issues but I'm not sure how popular generators are for uploads. Looking at the code for parallel upload that actually reads from this generator (where `get_chunk_streams()` is an async generator itself):
https://github.com/Azure/azure-sdk-for-python/blob/ec68720d53779dc97e16c39474f53fe41afe47bf/sdk/storage/azure-storage-blob/azure/storage/blob/_shared/uploads_async.py#L86-L96

https://github.com/Azure/azure-sdk-for-python/blob/ec68720d53779dc97e16c39474f53fe41afe47bf/sdk/storage/azure-storage-blob/azure/storage/blob/_shared/uploads_async.py#L22-L33

It seems we form the initial set of tasks synchronously but then will grab a new chunk after each task finishes. I'm not sure how async iterables work in terms of thread safety so I can't say for sure if this is thread safe or not. Thoughts?

Reply:
I added a test for this and ran it a number of times without issue but, I don't think that's enough to guarantee it works all the time.

Reply:
Discussed further offline and we think this should be okay.